File: native-host/host.mjs
#!/usr/bin/env node
// native-host/host.mjs
// MCP-Browser-Bridge native messaging host (Node.js, single file, zero npm deps).
//
// Bridges desktop MCP clients (HTTP / SSE on 127.0.0.1:7777) to the Chrome
// extension via Chrome Native Messaging (stdio with 4-byte LE length prefix).
//
// Pure Node.js built-in modules only: http, crypto, process.
// No Python. No npm install. Just run: node host.mjs
//
// Transport support (universal MCP client compatibility):
//   - MCP 2025-06-18 Streamable HTTP  : POST /mcp, GET /mcp, DELETE /mcp
//   - MCP 2024-11-05 legacy SSE       : GET /sse, POST /messages?sessionId=...
//
// The extension is the actual MCP server (registers tools, handles JSON-RPC).
// This host is just a transport adapter: HTTP client -> stdin -> extension ->
// stdout -> HTTP response.

import http from "node:http";
import crypto from "node:crypto";

// --------------------------------------------------------------------------- //
// Configuration
// --------------------------------------------------------------------------- //

const HOST = "127.0.0.1";
const DEFAULT_PORT = 7777;
const PORT = Number(process.env.MCPBB_PORT) || DEFAULT_PORT;

// Each MCP session has a random id and holds pending JSON-RPC responses keyed
// by jsonrpc id, plus a long-lived SSE response stream (for legacy SSE).
const sessions = new Map(); // sessionId -> { sseRes, pending: Map<id, {resolve,reject}>, streamInit: bool }

// Pending JSON-RPC requests sent to the extension but not yet answered,
// keyed by `${sessionId}:${jsonrpcId}`. Value is { resolve, reject, timer }.
const inflight = new Map();

// --------------------------------------------------------------------------- //
// Native messaging I/O (stdio)
// --------------------------------------------------------------------------- //
//
// Chrome native messaging protocol:
//   - stdin  : Chrome -> host. 4-byte LE length prefix + UTF-8 JSON.
//   - stdout : host -> Chrome. Same framing.
//   - stderr : free-form, captured by Chrome for debugging logs.

let stdinBuf = Buffer.alloc(0);

function sendToExtension(obj) {
  const json = Buffer.from(JSON.stringify(obj), "utf8");
  const header = Buffer.alloc(4);
  header.writeUInt32LE(json.length, 0);
  process.stdout.write(Buffer.concat([header, json]));
}

function logToStderr(msg) {
  process.stderr.write(`[mcpbb-host] ${msg}\n`);
}

process.stdin.on("data", (chunk) => {
  stdinBuf = Buffer.concat([stdinBuf, chunk]);
  while (stdinBuf.length >= 4) {
    const len = stdinBuf.readUInt32LE(0);
    if (stdinBuf.length < 4 + len) break;
    const json = stdinBuf.subarray(4, 4 + len);
    stdinBuf = stdinBuf.subarray(4 + len);
    let msg;
    try {
      msg = JSON.parse(json.toString("utf8"));
    } catch (e) {
      logToStderr(`invalid JSON from extension: ${e.message}`);
      continue;
    }
    handleExtensionMessage(msg).catch((e) => {
      logToStderr(`handleExtensionMessage error: ${e && e.message ? e.message : String(e)}`);
    });
  }
});

process.stdin.on("end", () => {
  logToStderr("stdin ended (Chrome disconnected). Shutting down.");
  shutdown(0);
});

// --------------------------------------------------------------------------- //
// Routing: messages coming back from the extension
// --------------------------------------------------------------------------- //
//
// The extension sends messages of the form:
//   { type: "response", clientId, payload: <json-rpc response> }
//   { type: "notification", clientId, payload: <json-rpc notification> }
//   { type: "bridge_ready", version, tools }
//   { type: "bridge_error", error }
//
// `clientId` is the MCP session id we assigned when forwarding the request.
// For Streamable HTTP, we resolve the pending inflight promise.
// For legacy SSE, we also write the response/notification to the SSE stream.

async function handleExtensionMessage(msg) {
  if (!msg || typeof msg !== "object") return;
  switch (msg.type) {
    case "response": {
      const { clientId, payload } = msg;
      deliverToSession(clientId, payload, /*isResponse=*/ true);
      break;
    }
    case "notification": {
      const { clientId, payload } = msg;
      deliverToSession(clientId, payload, /*isResponse=*/ false);
      break;
    }
    case "bridge_ready":
      logToStderr(`bridge ready: version=${msg.version} tools=${(msg.tools || []).length}`);
      break;
    case "bridge_error":
      logToStderr(`bridge error: ${msg.error}`);
      break;
    default:
      // Unknown: ignore silently.
      break;
  }
}

function deliverToSession(sessionId, payload, isResponse) {
  const session = sessions.get(sessionId);
  if (!session) {
    // Session already closed. Drop.
    return;
  }

  // Streamable HTTP: resolve inflight request by jsonrpc id.
  if (isResponse && payload && payload.id !== undefined && payload.id !== null) {
    const key = `${sessionId}:${payload.id}`;
    const entry = inflight.get(key);
    if (entry) {
      clearTimeout(entry.timer);
      inflight.delete(key);
      entry.resolve(payload);
    }
  }

  // Legacy SSE: also push the raw JSON-RPC message to the long-lived stream.
  if (session.sseRes && !session.sseRes.writableEnded) {
    const data = `data: ${JSON.stringify(payload)}\n\n`;
    session.sseRes.write(data);
    session.lastEventId = payload.id;
  }

  // If it was a response with an id and no SSE client is attached, the inflight
  // promise above already handled it. If there's no inflight (e.g. server-initiated
  // notification or a response we didn't track), and there's no SSE stream, just drop.
}

// --------------------------------------------------------------------------- //
// HTTP server: MCP 2025-06-18 Streamable HTTP + legacy SSE
// --------------------------------------------------------------------------- //

function newSessionId() {
  return crypto.randomBytes(16).toString("hex");
}

function ensureSession(id) {
  if (!sessions.has(id)) {
    sessions.set(id, { sseRes: null, lastEventId: null, pending: new Map(), streamInit: false });
  }
  return sessions.get(id);
}

function destroySession(id) {
  const s = sessions.get(id);
  if (!s) return;
  if (s.sseRes && !s.sseRes.writableEnded) {
    try { s.sseRes.end(); } catch (_) {}
  }
  // Reject any pending inflight requests on this session.
  for (const [key, entry] of inflight.entries()) {
    if (key.startsWith(`${id}:`)) {
      clearTimeout(entry.timer);
      entry.reject(new Error("session closed"));
      inflight.delete(key);
    }
  }
  sessions.delete(id);
}

// Forward a JSON-RPC payload to the extension tagged with the session id.
function forwardToExtension(sessionId, payload) {
  sendToExtension({ type: "request", clientId: sessionId, payload });
}

const server = http.createServer((req, res) => {
  // CORS: allow any local MCP client to call us.
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, POST, DELETE, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type, Accept, Mcp-Session-Id, Last-Event-ID");
  res.setHeader("Access-Control-Expose-Headers", "Mcp-Session-Id");
  if (req.method === "OPTIONS") {
    res.writeHead(204);
    res.end();
    return;
  }

  const url = new URL(req.url, `http://${HOST}:${PORT}`);
  const path = url.pathname;

  // ----------------------------------------------------------------------- //
  // Health / info endpoint
  // ----------------------------------------------------------------------- //
  if (path === "/" || path === "/health") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({
      name: "mcp-browser-bridge",
      version: "0.6.0",
      transport: ["streamable-http", "sse"],
      endpoints: {
        streamableHttp: "/mcp",
        sse: "/sse",
        messages: "/messages",
      },
    }));
    return;
  }

  // ----------------------------------------------------------------------- //
  // MCP 2025-06-18 Streamable HTTP: /mcp
  // ----------------------------------------------------------------------- //
  if (path === "/mcp") {
    handleStreamableHttp(req, res, url);
    return;
  }

  // ----------------------------------------------------------------------- //
  // MCP 2024-11-05 legacy SSE: /sse  +  /messages
  // ----------------------------------------------------------------------- //
  if (path === "/sse") {
    handleLegacySseHandshake(req, res, url);
    return;
  }
  if (path === "/messages") {
    handleLegacyMessages(req, res, url);
    return;
  }

  res.writeHead(404, { "Content-Type": "application/json" });
  res.end(JSON.stringify({ error: "not found", path }));
});

// --------------------------------------------------------------------------- //
// Streamable HTTP (2025-06-18)
// --------------------------------------------------------------------------- //
//
//   POST   /mcp  - client -> server JSON-RPC request/notification
//   GET    /mcp  - open a streaming response channel (server -> client)
//   DELETE /mcp  - terminate the session

function handleStreamableHttp(req, res, url) {
  const accept = req.headers["accept"] || "";
  const sessionHeader = req.headers["mcp-session-id"];

  if (req.method === "POST") {
    let body = "";
    req.on("data", (c) => { body += c; });
    req.on("end", () => {
      let payload;
      try {
        payload = JSON.parse(body);
      } catch (e) {
        res.writeHead(400, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ jsonrpc: "2.0", error: { code: -32700, message: "Parse error" }, id: null }));
        return;
      }

      // Assign or reuse a session.
      const sessionId = sessionHeader || newSessionId();
      ensureSession(sessionId);

      // Notifications (no id): forward, acknowledge with 202.
      if (payload && (payload.id === undefined || payload.id === null)) {
        forwardToExtension(sessionId, payload);
        res.writeHead(202, {
          "Content-Type": "application/json",
          "Mcp-Session-Id": sessionId,
        });
        res.end(JSON.stringify({ ok: true }));
        return;
      }

      // Request (has id): forward, wait for response, return it inline.
      const key = `${sessionId}:${payload.id}`;
      const timer = setTimeout(() => {
        const entry = inflight.get(key);
        if (entry) {
          inflight.delete(key);
          entry.reject(new Error("timeout waiting for extension response"));
        }
      }, 5 * 60 * 1000); // 5 min timeout

      inflight.set(key, {
        resolve: (response) => {
          res.writeHead(200, {
            "Content-Type": "application/json",
            "Mcp-Session-Id": sessionId,
          });
          res.end(JSON.stringify(response));
        },
        reject: (err) => {
          res.writeHead(500, {
            "Content-Type": "application/json",
            "Mcp-Session-Id": sessionId,
          });
          res.end(JSON.stringify({
            jsonrpc: "2.0",
            error: { code: -32603, message: err.message || "internal error" },
            id: payload.id,
          }));
        },
        timer,
      });

      forwardToExtension(sessionId, payload);
    });
    req.on("error", (e) => {
      logToStderr(`POST /mcp request error: ${e.message}`);
      try { res.writeHead(400); res.end(); } catch (_) {}
    });
    return;
  }

  if (req.method === "GET") {
    // Open a streaming response channel. The client uses Accept: text/event-stream.
    if (!accept.includes("text/event-stream")) {
      res.writeHead(406, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ error: "GET /mcp requires Accept: text/event-stream" }));
      return;
    }
    const sessionId = sessionHeader || newSessionId();
    const session = ensureSession(sessionId);
    res.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      "Connection": "keep-alive",
      "Mcp-Session-Id": sessionId,
    });
    // Keep the stream open; server-initiated notifications will be pushed here.
    // We don't yet expose server-initiated notifications from the extension, but
    // the channel is ready if/when we add them.
    res.write(`event: endpoint\ndata: ${JSON.stringify("/mcp")}\n\n`);
    // Note: we intentionally do NOT attach this as session.sseRes because the
    // Streamable HTTP model resolves each POST inline. This GET channel is for
    // server-initiated notifications only (rare). Keep a reference so we can
    // close it on session destroy.
    session.streamRes = res;
    req.on("close", () => {
      if (session.streamRes === res) session.streamRes = null;
    });
    return;
  }

  if (req.method === "DELETE") {
    const sessionId = sessionHeader;
    if (sessionId) destroySession(sessionId);
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ ok: true }));
    return;
  }

  res.writeHead(405, { "Content-Type": "application/json" });
  res.end(JSON.stringify({ error: "method not allowed" }));
}

// --------------------------------------------------------------------------- //
// Legacy SSE (2024-11-05)
// --------------------------------------------------------------------------- //
//
//   GET  /sse                         - open SSE stream, receive endpoint URI
//   POST /messages?sessionId=...      - send JSON-RPC to the server

function handleLegacySseHandshake(req, res, url) {
  // In the legacy protocol, the session id is chosen by the server and
  // embedded in the endpoint URI returned on the SSE stream.
  const sessionId = newSessionId();
  const session = ensureSession(sessionId);
  session.sseRes = res;

  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
  });
  const endpointUri = `/messages?sessionId=${sessionId}`;
  res.write(`event: endpoint\ndata: ${JSON.stringify(endpointUri)}\n\n`);

  // Heartbeat every 25s to keep proxies from closing the connection.
  const heartbeat = setInterval(() => {
    try { res.write(`: ping\n\n`); } catch (_) {}
  }, 25000);
  session.heartbeat = heartbeat;

  req.on("close", () => {
    clearInterval(heartbeat);
    if (session.sseRes === res) session.sseRes = null;
    // Note: we do not destroy the session here; the client may reconnect.
    // The session will be reaped if unused.
  });
}

function handleLegacyMessages(req, res, url) {
  const sessionId = url.searchParams.get("sessionId");
  if (!sessionId) {
    res.writeHead(400, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ error: "missing sessionId query parameter" }));
    return;
  }
  const session = sessions.get(sessionId);
  if (!session) {
    res.writeHead(404, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ error: "unknown session" }));
    return;
  }

  if (req.method !== "POST") {
    res.writeHead(405, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ error: "POST only" }));
    return;
  }

  let body = "";
  req.on("data", (c) => { body += c; });
  req.on("end", () => {
    let payload;
    try {
      payload = JSON.parse(body);
    } catch (e) {
      res.writeHead(400, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ jsonrpc: "2.0", error: { code: -32700, message: "Parse error" }, id: null }));
      return;
    }

    // In the legacy SSE protocol, POST /messages is acknowledged with 202 and
    // the actual JSON-RPC response is delivered via the SSE stream.
    res.writeHead(202, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ ok: true }));

    forwardToExtension(sessionId, payload);
  });
  req.on("error", (e) => {
    logToStderr(`POST /messages request error: ${e.message}`);
    try { res.writeHead(400); res.end(); } catch (_) {}
  });
}

// --------------------------------------------------------------------------- //
// Startup / shutdown
// --------------------------------------------------------------------------- //

server.on("error", (err) => {
  if (err.code === "EADDRINUSE") {
    logToStderr(`port ${PORT} is already in use. Set MCPBB_PORT to override.`);
  } else {
    logToStderr(`http server error: ${err.message}`);
  }
  shutdown(1);
});

server.listen(PORT, HOST, () => {
  logToStderr(`MCP-Browser-Bridge native host listening on http://${HOST}:${PORT}`);
  logToStderr(`  Streamable HTTP : http://${HOST}:${PORT}/mcp`);
  logToStderr(`  Legacy SSE      : http://${HOST}:${PORT}/sse`);
  logToStderr(`  Health          : http://${HOST}:${PORT}/health`);
  // Announce to the extension that the host is up.
  sendToExtension({ type: "host_ready", host: HOST, port: PORT, version: "0.6.0" });
});

let shuttingDown = false;
function shutdown(code) {
  if (shuttingDown) return;
  shuttingDown = true;
  try { server.close(); } catch (_) {}
  for (const id of [...sessions.keys()]) destroySession(id);
  process.exit(code);
}

process.on("SIGINT", () => shutdown(0));
process.on("SIGTERM", () => shutdown(0));
process.on("uncaughtException", (e) => {
  logToStderr(`uncaughtException: ${e && e.stack ? e.stack : String(e)}`);
});
process.on("unhandledRejection", (e) => {
  logToStderr(`unhandledRejection: ${e && e.message ? e.message : String(e)}`);
});

File: native-host/host.py
#!/usr/bin/env python3
"""
MCP-Browser-Bridge native messaging host.

Role: a thin, fast I/O proxy between MCP clients (Claude Desktop, etc.) and
the browser extension. It owns the local HTTP/SSE server socket and forwards
raw JSON-RPC frames to the extension over Chrome native messaging (stdio).
All MCP protocol parsing happens inside the extension; this host never
inspects or transforms payloads, which keeps latency minimal.

Transport:
  - SSE:   GET  /sse                       -> event stream (response + endpoint event)
           POST /messages?sessionId=<id>    -> send JSON-RPC, response via stream
  - HTTP:  POST /mcp                       -> single request/response (synchronous)
  - both:  both endpoints are served

Message flow:
  MCP client --HTTP/SSE--> host --native msg--> extension (parses + handles)
  extension --native msg--> host --HTTP/SSE--> MCP client

Native messaging wire format (Chrome spec):
  stdin/stdout messages are: <uint32 little-endian length><utf-8 json bytes>
"""
import sys
import os
import json
import struct
import threading
import queue
import uuid
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs

HOST_NAME = "com.mcpbrowser.bridge"


# --------------------------------------------------------------------------- #
# Native messaging I/O (thread-safe writes; single stdin reader)
# --------------------------------------------------------------------------- #
_write_lock = threading.Lock()


def _log(msg):
    """Safe stderr write (Windows pythonw has stderr=None)."""
    try:
        if sys.stderr is not None:
            sys.stderr.write(msg)
            sys.stderr.flush()
    except Exception:
        pass


def read_message():
    """Read one native messaging message from stdin. Returns dict or None on EOF."""
    raw_len = sys.stdin.buffer.read(4)
    if len(raw_len) < 4:
        return None
    n = struct.unpack("<I", raw_len)[0]
    if n <= 0 or n > 64 * 1024 * 1024:  # Chrome caps messages at 1 MB; be lenient
        return None
    data = b""
    while len(data) < n:
        chunk = sys.stdin.buffer.read(n - len(data))
        if not chunk:
            return None
        data += chunk
    try:
        return json.loads(data.decode("utf-8"))
    except Exception as e:
        _log(f"[host] bad json from extension: {e}\n")
        return None


def write_message(msg):
    """Write one native messaging message to stdout (thread-safe)."""
    data = json.dumps(msg, separators=(",", ":")).encode("utf-8")
    with _write_lock:
        try:
            sys.stdout.buffer.write(struct.pack("<I", len(data)))
            sys.stdout.buffer.write(data)
            sys.stdout.buffer.flush()
        except Exception:
            # stdout closed (browser is shutting down); ignore.
            pass


# --------------------------------------------------------------------------- #
# Bridge state
# --------------------------------------------------------------------------- #
class Client:
    __slots__ = ("client_id", "transport", "queue", "alive")

    def __init__(self, client_id, transport):
        self.client_id = client_id
        self.transport = transport  # "sse" | "http"
        self.queue = queue.Queue()  # outgoing ("response"|"notification", payload)
        self.alive = True


class BridgeHost:
    def __init__(self):
        self.clients = {}
        self.lock = threading.Lock()
        self.server = None
        self.server_thread = None
        self.config = {"host": "127.0.0.1", "port": 8765, "transport": "sse"}
        self.running = True

    # --- client registry ---
    def add_client(self, client_id, transport):
        with self.lock:
            self.clients[client_id] = Client(client_id, transport)
            return self.clients[client_id]

    def get_client(self, client_id):
        with self.lock:
            return self.clients.get(client_id)

    def remove_client(self, client_id):
        with self.lock:
            self.clients.pop(client_id, None)

    # --- extension message handling ---
    def handle_extension_message(self, msg):
        t = msg.get("type")
        if t == "reconfigure":
            self.reconfigure(msg)
        elif t == "response":
            self._route(msg.get("clientId"), ("response", msg.get("payload")))
        elif t == "notification":
            self._route(msg.get("clientId"), ("notification", msg.get("payload")))
        elif t == "ping":
            write_message({"type": "pong", "ts": int(time.time() * 1000)})
        # unknown types: ignore

    def _route(self, client_id, item):
        if not client_id:
            return
        c = self.get_client(client_id)
        if c:
            c.queue.put(item)

    def reconfigure(self, msg):
        new_cfg = {
            "host": msg.get("host", self.config["host"]),
            "port": int(msg.get("port", self.config["port"])),
            "transport": msg.get("transport", self.config["transport"]),
        }
        same = (
            self.server is not None
            and new_cfg["host"] == self.config["host"]
            and new_cfg["port"] == self.config["port"]
            and new_cfg["transport"] == self.config["transport"]
        )
        if same:
            write_message({"type": "config", **new_cfg, "already_running": True})
            return
        self.config = new_cfg
        self.stop_server()
        ok = self.start_server()
        write_message({
            "type": "config",
            **new_cfg,
            "started": ok,
        })

    def start_server(self):
        host, port, transport = self.config["host"], self.config["port"], self.config["transport"]
        try:
            srv = BridgeServer((host, port), self)
            self.server = srv
            self.server_thread = threading.Thread(target=srv.serve_forever, daemon=True)
            self.server_thread.start()
            _log(f"[host] listening on {host}:{port} transport={transport}\n")
            return True
        except Exception as e:
            _log(f"[host] failed to start server: {e}\n")
            self.server = None
            return False

    def stop_server(self):
        if self.server is None:
            return
        try:
            # shutdown must be called from another thread to avoid deadlock
            threading.Thread(target=self.server.shutdown, daemon=True).start()
            self.server.server_close()
        except Exception:
            pass
        self.server = None

    def run(self):
        """Main loop: read from extension stdin until EOF."""
        try:
            while self.running:
                msg = read_message()
                if msg is None:
                    break
                try:
                    self.handle_extension_message(msg)
                except Exception as e:
                    _log(f"[host] error handling message: {e}\n")
        finally:
            self.running = False
            self.stop_server()


# --------------------------------------------------------------------------- #
# HTTP/SSE server
# --------------------------------------------------------------------------- #
class BridgeServer(ThreadingHTTPServer):
    daemon_threads = True
    allow_reuse_address = True

    def __init__(self, server_address, bridge):
        super().__init__(server_address, Handler)
        self.bridge = bridge


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    server_version = "MCPBrowserBridge/0.1"

    # quieter logging
    def log_message(self, *args):
        pass

    @property
    def bridge(self):
        return self.server.bridge

    @property
    def transport_mode(self):
        return self.bridge.config.get("transport", "sse")

    # ----- helpers -----
    def send_json(self, obj, status=200):
        body = json.dumps(obj).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(body)

    def read_json_body(self):
        length = int(self.headers.get("Content-Length", "0") or "0")
        body = self.rfile.read(length) if length > 0 else b""
        if not body:
            return {}
        return json.loads(body.decode("utf-8"))

    # ----- GET -----
    def do_GET(self):
        parsed = urlparse(self.path)
        path = parsed.path
        if path in ("/", "/health"):
            self.send_json({
                "status": "ok",
                "name": HOST_NAME,
                "transport": self.transport_mode,
                "endpoint": f"http://{self.bridge.config['host']}:{self.bridge.config['port']}",
            })
            return
        if path == "/sse":
            if self.transport_mode in ("sse", "both"):
                self.handle_sse()
            else:
                self.send_error(404, "SSE not enabled")
            return
        self.send_error(404)

    # ----- POST -----
    def do_POST(self):
        parsed = urlparse(self.path)
        path = parsed.path
        try:
            payload = self.read_json_body()
        except Exception:
            self.send_error(400, "invalid json")
            return

        if path == "/messages" and self.transport_mode in ("sse", "both"):
            self.handle_sse_message(parsed, payload)
            return
        if path == "/mcp" and self.transport_mode in ("http", "both"):
            self.handle_http_message(payload)
            return
        # Fallback: accept /messages for http mode and /mcp for sse mode too,
        # to be forgiving with clients.
        if path in ("/mcp", "/messages"):
            qs = parse_qs(parsed.query)
            session_id = qs.get("sessionId", [None])[0]
            if session_id and self.bridge.get_client(session_id):
                self.handle_sse_message(parsed, payload)
            else:
                self.handle_http_message(payload)
            return
        self.send_error(404)

    # ----- SSE -----
    def handle_sse(self):
        client_id = "sse-" + uuid.uuid4().hex
        self.bridge.add_client(client_id, "sse")
        write_message({"type": "client_connect", "clientId": client_id, "transport": "sse"})

        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-store, no-transform")
        self.send_header("Connection", "keep-alive")
        self.send_header("X-Accel-Buffering", "no")
        self.end_headers()

        # Tell the client where to POST subsequent requests.
        endpoint = f"/messages?sessionId={client_id}"
        try:
            self.wfile.write(f"event: endpoint\ndata: {endpoint}\n\n".encode("utf-8"))
            self.wfile.flush()
        except Exception:
            self._cleanup_sse(client_id)
            return

        client = self.bridge.get_client(client_id)
        last_ping = time.time()
        try:
            while self.bridge.running and client.alive:
                try:
                    kind, payload = client.queue.get(timeout=15)
                except queue.Empty:
                    # heartbeat
                    now = time.time()
                    if now - last_ping >= 15:
                        self.wfile.write(b": ping\n\n")
                        self.wfile.flush()
                        last_ping = now
                    continue
                data = json.dumps(payload, separators=(",", ":"))
                # SSE "data" lines must not contain raw newlines; split safely.
                safe = data.replace("\n", "\ndata: ")
                self.wfile.write(f"data: {safe}\n\n".encode("utf-8"))
                self.wfile.flush()
        except Exception:
            pass
        finally:
            self._cleanup_sse(client_id)

    def _cleanup_sse(self, client_id):
        self.bridge.remove_client(client_id)
        write_message({"type": "client_disconnect", "clientId": client_id})

    def handle_sse_message(self, parsed, payload):
        qs = parse_qs(parsed.query)
        session_id = qs.get("sessionId", [None])[0]
        client = self.bridge.get_client(session_id) if session_id else None
        if not client:
            # No active SSE session; fall back to synchronous HTTP response.
            self.handle_http_message(payload)
            return
        # Forward to extension; response will arrive over the SSE stream.
        write_message({"type": "request", "clientId": session_id, "payload": payload})
        # Acknowledge receipt. The actual JSON-RPC response is pushed via SSE.
        self.send_response(202)
        self.send_header("Content-Length", "0")
        self.end_headers()

    # ----- HTTP (synchronous) -----
    def handle_http_message(self, payload):
        client_id = "http-" + uuid.uuid4().hex
        self.bridge.add_client(client_id, "http")
        write_message({"type": "client_connect", "clientId": client_id, "transport": "http"})
        write_message({"type": "request", "clientId": client_id, "payload": payload})

        # Notifications have no id and expect no response.
        is_notification = isinstance(payload, dict) and "id" not in payload
        if is_notification:
            self.bridge.remove_client(client_id)
            write_message({"type": "client_disconnect", "clientId": client_id})
            self.send_response(202)
            self.send_header("Content-Length", "0")
            self.end_headers()
            return

        client = self.bridge.get_client(client_id)
        response = None
        try:
            kind, response = client.queue.get(timeout=30)
        except queue.Empty:
            response = {
                "jsonrpc": "2.0",
                "id": payload.get("id") if isinstance(payload, dict) else None,
                "error": {"code": -32603, "message": "timeout waiting for extension"},
            }
        finally:
            self.bridge.remove_client(client_id)
            write_message({"type": "client_disconnect", "clientId": client_id})

        self.send_json(response)


# --------------------------------------------------------------------------- #
# Entry point
# --------------------------------------------------------------------------- #
def main():
    # On Windows, Chrome launches hosts without a console; stderr is harmless.
    # On Linux/macOS, stderr is discarded by Chrome but useful for manual runs.
    _log("[host] MCP-Browser-Bridge native host starting\n")
    host = BridgeHost()
    host.run()
    _log("[host] native host exiting\n")


if __name__ == "__main__":
    main()
